### ISSUE TITLE
Protect people's passwords

### ISSUE THREAD

**Issue by Philipp91** [doctrine/dbal#2698]
I'm referring to the warning here: https://secure.php.net/manual/en/pdo.connections.php
If you know German, you can also read this: https://www.golem.de/news/it-sicherheit-wie-ich-mein-passwort-im-stack-trace-fand-1704-127258.html

A default PHP installation will leak your MySQL Password when connecting with Doctrine DBAL (or any other library or using PDO directly), if the connection fails (DB server unreachable, wrong username, etc.). Because `display_errors` is turned on by default, the stack trace will be printed, and it includes Doctrine's line https://github.com/doctrine/dbal/blob/master/lib/Doctrine/DBAL/Connection.php#L363 where the password is present, so it's printed out to the client in cleartext.

To prevent this, you can obviously do things like turn off `display_errors` or globally catch all errors. But that might not be in the hands of the developer writing code with PDO/Doctrine, so in order to protect users of your code, you have to catch the exception. The official PHP documentation linked above contains a warning that explains it. Normally, I do this myself when I create the connection.

But when I use Doctrine DBAL, it seems that the connection is only established once I actually make a query (which is generally a good thing). So the stack trace looks roughly like this:

```
1. in PDOConnection.php line 43
2. at PDO->__construct('mysql:host=localhost;dbname=sdfsdf;charset=utf8mb4;', 'root', 'sdfsdf2', array(true, 'SET time_zone = \'+00:00\'')) in PDOConnection.php line 43
3. at PDOConnection->__construct('mysql:host=localhost;dbname=sdfsdf;charset=utf8mb4;', 'root', 'sdfsdf2', array(true, 'SET time_zone = \'+00:00\'')) in Driver.php line 41
4. at Driver->connect(array('driver' => 'pdo_mysql', 'dbname' => 'sdfsdf', 'host' => 'localhost', 'user' => 'root', 'password' => 'sdfsdf2', 'charset' => 'utf8mb4', 'driverOptions' => array(true, 'SET time_zone = \'+00:00\'')), 'root', 'sdfsdf2', array(true, 'SET time_zone = \'+00:00\'')) in Connection.php line 360
5. at Connection->connect() in Connection.php line 814
6. at Connection->executeQuery('SELECT * FROM token WHERE token = ?', array('0be7567c24a8421e657c369a22357a7384270f92097fe9a76d9d6e1b2c0d8f47')) in TokenDAO.php line 42
```

Note that the password (changed to `'sdfsdf2'` here) occurs a couple of times. `TokenDAO` is my application's DB/DAO layer. So if an application using Doctrine DBAL wants to prevent such password leaks, the developer has to enclose **every** query call (like `executeQuery` above) in a try-catch, and must not miss one.

tl;dr
That's why I propose that, instead, Doctrine protects its users and wraps its `Driver::connect()` call here https://github.com/doctrine/dbal/blob/master/lib/Doctrine/DBAL/Connection.php#L363 in a try-catch. When an exception occurs, Doctrine could either throw a new one (just saying that the DB connection failed for some reason), or somehow modify and re-throw the existing one to remove the detailed stack trace, if that's possible at all.

I could try and submit a patch for that.

**Comment by Ocramius** [comment]
As long as your password is in any variable passed to any function literally *ANYWHERE* where you could have a failure, this is going to happen anyway. A try-catch is just a shotgun patch, and it would just cause more problems at debugging.

Closing as duplicate of #2337.

This has nothing to do with doctrine or your application, but rather with your configuration.

fix your logging/error reporting settings instead, as errors (any errors) can expose:
 * sensitive customer information
 * sensitive error information
 * sensitive environment information
 * possible security issues/unescaped input handling
